src/main/java/ch/njol/skript/expressions/ExprDirection.java

```diff
@@ -124,89 +125,104 @@ public boolean init(final Expression<?>[] exprs, final int matchedPattern, final
+						return CollectionUtils.array(new Direction(new double[] {0, -length, 0}));
 				}
-				double yaw = Direction.yawToRadians(l.getYaw());
+				double yaw = Direction.yawToRadians(location.getYaw());
 				if (horizontal && !facing) {
-					return new Direction[] {new Direction(new double[] {Math.cos(yaw) * ln, 0, Math.sin(yaw) * ln})};
+					return CollectionUtils.array(new Direction(new double[] {Math.cos(yaw) * length, 0, Math.sin(yaw) * length}));
 				}
 				yaw = Math2.mod(yaw, 2 * Math.PI);
 				if (yaw >= Math.PI / 4 && yaw < 3 * Math.PI / 4)
-					return new Direction[] {new Direction(new double[] {0, 0, ln})};
+					return CollectionUtils.array(new Direction(new double[] {0, 0, length}));
 				if (yaw >= 3 * Math.PI / 4 && yaw < 5 * Math.PI / 4)
-					return new Direction[] {new Direction(new double[] {-ln, 0, 0})};
+					return CollectionUtils.array(new Direction(new double[] {-length, 0, 0}));
 				if (yaw >= 5 * Math.PI / 4 && yaw < 7 * Math.PI / 4)
-					return new Direction[] {new Direction(new double[] {0, 0, -ln})};
+					return CollectionUtils.array(new Direction(new double[] {0, 0, -length}));
 				assert yaw >= 0 && yaw < Math.PI / 4 || yaw >= 7 * Math.PI / 4 && yaw < 2 * Math.PI;
-				return new Direction[] {new Direction(new double[] {ln, 0, 0})};
+				return CollectionUtils.array(new Direction(new double[] {length, 0, 0}));
+			} else if (object instanceof Location) {
+				assert this.from != null;
+				Location from = this.from.getSingle(event);
+				if (from == null)
+					return new Direction[0];
+				Location to = (Location) object;
+				return CollectionUtils.array(new Direction(to.toVector().subtract(from.toVector())));
 			}
+			assert false;
+			return new Direction[0];
 		} else {
-			return new Direction[] {new Direction(horizontal ? Direction.IGNORE_PITCH : 0, yaw, ln)};
+			return CollectionUtils.array(new Direction(horizontal ? Direction.IGNORE_PITCH : 0, yaw, length));
 		}
 	}
-	
+
 	@Override
 	public boolean isSingle() {
 		return true;
 	}
-	
+
 	@Override
 	public Class<? extends Direction> getReturnType() {
 		return Direction.class;
 	}
-	
+
 	@Override
-	public String toString(final @Nullable Event e, final boolean debug) {
-		final Expression<?> relativeTo = this.relativeTo;
-		return (amount != null ? amount.toString(e, debug) + " meter(s) " : "") + (direction != null ? Direction.toString(direction) :
-				relativeTo != null ? " in " + (horizontal ? "horizontal " : "") + (facing ? "facing" : "direction") + " of " + relativeTo.toString(e, debug) :
-						(horizontal ? "horizontally " : "") + Direction.toString(0, yaw, 1));
+	public String toString(@Nullable Event event, boolean debug) {
+		return (amount != null ? amount.toString(event, debug) + " meter(s) " : "") +
+				(direction != null ? Direction.toString(direction) : relativeTo != null ? " in " +
+				(horizontal ? "horizontal " : "") +(facing ? "facing" : "direction") + " of " +
```

Comment:
Would this be a good opportunity to clean up the giant mess of 6 nested ternary statements?(:?:?:?:)

Reply:
Yeah could use the SyntaxStringBuilder